pkg/macaroon/serialize.go

```diff
@@ -0,0 +1,205 @@
+		section = section[1:]
+	}
+	if len(section) != 1 || section[0].fieldType != fieldIdentifier {
+		return nil, errors.New("invalid macaroon header")
+	}
+
+	mac := Macaroon{}
+	mac.head = section[0].data
+	for {
+		rest, section, err := parseSection(data)
+		if err != nil {
+			return nil, err
+		}
+		data = rest
+		if len(section) == 0 {
+			break
+		}
+		if len(section) > 0 && section[0].fieldType == fieldLocation {
+			section = section[1:]
+		}
+		if len(section) == 0 || section[0].fieldType != fieldIdentifier {
+			return nil, errors.New("no Identifier in caveat")
+		}
+		cav := append([]byte(nil), section[0].data...)
+		section = section[1:]
+		if len(section) == 0 {
+			// First party caveat.
+			//if cav.Location != "" {
+			//	return nil, errors.New("location not allowed in first party caveat")
+			//}
+			mac.caveats = append(mac.caveats, cav)
+			continue
+		}
+		if len(section) != 1 {
+			return nil, errors.New("extra fields found in caveat")
+		}
+		if section[0].fieldType != fieldVerificationID {
+			return nil, errors.New("invalid field found in caveat")
+		}
+		//cav.VerificationId = section[0].data
+		mac.caveats = append(mac.caveats, cav)
+	}
+	data, sig, err := parsePacket(data)
+	if err != nil {
+		return nil, err
+	}
+	if sig.fieldType != fieldSignature {
+		return nil, errors.New("unexpected field found instead of signature")
+	}
+	if len(sig.data) != 32 {
+		return nil, errors.New("signature has unexpected length")
+	}
+	mac.tail = make([]byte, 32)
+	copy(mac.tail[:], sig.data)
+	//return data, nil
+	// Parse Identity
+	// Parse caveats
+	// Parse tail
+	return &mac, nil
+}
+
+// parseSection returns data leftover and packet array
+func parseSection(data []byte) ([]byte, []packet, error) {
+	prevFieldType := fieldType(-1)
+	var packets []packet
+	for {
+		if len(data) == 0 {
+			return nil, nil, errors.New("section extends past end of buffer")
+		}
+		rest, p, err := parsePacket(data)
+		if err != nil {
+			return nil, nil, err
+		}
+		if p.fieldType == fieldEOS {
+			return rest, packets, nil
+		}
+		if p.fieldType <= prevFieldType {
+			return nil, nil, errors.New("fields out of order")
+		}
+		packets = append(packets, p)
+		prevFieldType = p.fieldType
+		data = rest
+	}
+}
+
+// parsePacket returns data leftover and packet
+func parsePacket(data []byte) ([]byte, packet, error) {
+	data, ft, err := parseVarint(data)
+	if err != nil {
+		return nil, packet{}, err
+	}
+
+	p := packet{fieldType: fieldType(ft)}
+	if p.fieldType == fieldEOS {
+		return data, p, nil
+	}
+	data, packLen, err := parseVarint(data)
+	if err != nil {
+		return nil, packet{}, err
+	}
+
+	if packLen > len(data) {
+		return nil, packet{}, errors.New("out of bounds")
+	}
+	if packLen == 0 {
+		p.data = nil
+
+		return data, p, nil
+	}
+
+	p.data = data[0:packLen]
+
+	return data[packLen:], p, nil
+}
+
+func parseVarint(data []byte) ([]byte, int, error) {
+	value, n := binary.Uvarint(data)
+	if n <= 0 || value > 0x7fffffff {
```

Comment:
if I'm not mistaken, `0x7fffffff` is the maximum value that an int32 can hold.
Being that the case, use `Math.MaxInt32` because it's clearer for the reader.